src/zenml/integrations/kubernetes/template_engine.py

```diff
@@ -0,0 +1,572 @@
+
+        Raises:
+            ValueError: If a remote path is provided or objects cannot be serialized.
+        """
+        if output_dir is not None:
+            output_dir_str = str(output_dir)
+            if io_utils.is_remote(output_dir_str):
+                raise ValueError(
+                    f"Remote paths are not supported: {output_dir_str}"
+                )
+            base_dir = io_utils.resolve_relative_path(output_dir_str)
+            manifest_dir = os.path.join(base_dir, deployment_name)
+        else:
+            manifest_dir = (
+                KubernetesTemplateEngine._determine_default_manifest_dir(
+                    deployment_name
+                )
+            )
+
+        os.makedirs(manifest_dir, exist_ok=True)
+
+        saved_files = []
+        for idx, k8s_obj in enumerate(k8s_objects):
+            if not hasattr(k8s_obj, "to_dict"):
+                raise ValueError(
+                    f"Object at index {idx} does not have a to_dict() method"
+                )
+            obj_dict = k8s_obj.to_dict()
+
+            obj_dict = self._clean_k8s_resource_dict(obj_dict)
+
+            kind = obj_dict.get("kind", "unknown").lower()
+            name = obj_dict.get("metadata", {}).get("name", f"resource-{idx}")
+            filename = f"{kind}-{name}.yaml"
+            filepath = os.path.join(manifest_dir, filename)
+
+            with open(filepath, "w") as f:
+                yaml.dump(
+                    obj_dict, f, default_flow_style=False, sort_keys=False
+                )
+
+            saved_files.append(filepath)
+
+        logger.info(
+            f"Saved {len(saved_files)} validated Kubernetes manifest(s) to {manifest_dir}"
+        )
+        return saved_files
+
+    @staticmethod
+    def _clean_k8s_resource_dict(obj_dict: Dict[str, Any]) -> Dict[str, Any]:
+        """Remove Kubernetes internal metadata for clean YAML output.
+
+        Args:
+            obj_dict: Raw Kubernetes resource dictionary.
+
+        Returns:
+            Cleaned dictionary with only user-relevant fields.
+        """
+        obj_dict.pop("status", None)
+
+        metadata = obj_dict.get("metadata", {})
+        if metadata:
+            for field in [
+                "managedFields",
+                "uid",
+                "resourceVersion",
+                "generation",
+                "creationTimestamp",
+                "selfLink",
+            ]:
+                metadata.pop(field, None)
+
+            obj_dict["metadata"] = {
+                k: v
+                for k, v in metadata.items()
+                if v not in (None, "", {}, [])
+            }
+
+        return obj_dict
+
+    # ========================================================================
+    # Template and Resource Rendering
+    # ========================================================================
+
+    def render_template(
+        self,
+        template_or_file: str,
```

Comment:
To avoid confusion and possible mistakes, can you please make this function only take `template_file` as input.